### PR TITLE
EU-105

### PR DESCRIPTION
**Pull request by YehorVoytenko**
https://euniversity.atlassian.net/jira/software/projects/EU/boards/1?selectedIssue=EU-105

### CODE REVIEW

File: apps/services/schemas.py [hunk 1]
```diff
@@ -118,3 +118,13 @@ class UserRequestDetailsViewOut(BaseModel):
     date_review: datetime = None
     remark: str = None
     documents: List[Dict[str, str]]
+
+
+class CountHostelAccommodationIn(BaseModel):
```
Comment (anton-serozhechkin) [inline apps/services/schemas.py]: CountHostelAccommodationCostIn is better

File: apps/services/schemas.py [hunk 2]
```diff
@@ -118,3 +118,13 @@ class UserRequestDetailsViewOut(BaseModel):
...
+    start_date_accommodation: date = None
+    end_date_accommodation: date = None
+    bed_place_id: int = None
+
+
+class CountHostelAccommodationOut(BaseModel):
```
Comment (anton-serozhechkin) [inline apps/services/schemas.py]: CountHostelAccommodationCostOut is better

File: apps/services/schemas.py [hunk 3]
```diff
@@ -118,3 +118,13 @@ class UserRequestDetailsViewOut(BaseModel):
     date_review: datetime = None
     remark: str = None
     documents: List[Dict[str, str]]
+
+
+class CountHostelAccommodationIn(BaseModel):
+    start_date_accommodation: date = None
```
Comment (anton-serozhechkin) [inline apps/services/schemas.py]: all these 3 fields are required, please delete " = None"

File: apps/services/schemas.py [hunk 4]
```diff
@@ -118,3 +118,13 @@ class UserRequestDetailsViewOut(BaseModel):
...
+    end_date_accommodation: date = None
+    bed_place_id: int = None
+
+
+class CountHostelAccommodationOut(BaseModel):
+    total_accommodation_cost: float
```
Comment (anton-serozhechkin) [inline apps/services/schemas.py]: change type to decimal and rename to total_hostel_accommodation_cost

File: apps/services/services.py [hunk 5]
```diff
@@ -40,3 +40,13 @@ async def create_user_document(**kwargs):
                                         content=content,
                                         user_request_id=kwargs.get("user_request_id"))
     return await database.execute(query)
+
+
+def calculate_accommodation(end_date, start_date, month_price, bed_place_name):
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: calculate_hostel_accommodation_cost is better
also please add hints for params and for result of method, i mean their types

File: apps/services/services.py [hunk 6]
```diff
@@ -40,3 +40,13 @@ async def create_user_document(**kwargs):
...
+    if end_date <= start_date:
+        raise ValueError(
+            message="Wrong date value",
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: It's better to describe error more deeply. For example:  Start date of hostel accommodation can't be more or equal than end date of hostel accomodation

File: apps/services/services.py [hunk 7]
```diff
@@ -40,3 +40,13 @@ async def create_user_document(**kwargs):
                                         content=content,
                                         user_request_id=kwargs.get("user_request_id"))
     return await database.execute(query)
+
+
+def calculate_accommodation(end_date, start_date, month_price, bed_place_name):
+    if end_date <= start_date:
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: i think it's better to move the checking of dates to "In" schema, to validate it before triggering the endpoint

File: apps/services/handlers.py [hunk 8]
```diff
@@ -240,11 +241,11 @@ async def create_user_request_review(university_id: int, user_request_id: int, u
 
 @services_router.get("/{university_id}/hostel-accommodation/{user_request_id}",
                      name="read_hostel_accommodation",
-                     response_model=JSENDOutSchema[HostelAccomodationViewOut],
+                     response_model=JSENDOutSchema[HostelAccommodationViewOut],
                      summary="Get hostel accommodation",
                      responses={200: {"description": "Successful get user request hostel accommodation response"}},
                      tags=["Student dashboard"])
-async def read_hostel_accommodation(university_id: int, user_request_id: int, user=Depends(get_current_user)):
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: why did you delete it?

File: apps/services/handlers.py [hunk 9]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
         "data": response,
         "message": "Got request details"
     }
+
+
+@services_router.post("/{user_request_id}/student-accommodation/{university_id}/",
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: 1. first param in URL is always must be university_id
2. please rename URL to count-hostel-accommodation-cost
3. we don't need to receive user_request_id, because endpoint must count the cost of  hostel accommodation despite user request

File: apps/services/handlers.py [hunk 10]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+                      name="count_student_accommodation",
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: create_count_hostel_accommodation_cost is better

File: apps/services/handlers.py [hunk 11]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+                      response_model=JSENDOutSchema[CountHostelAccommodationIn],
+                      tags=["Student dashboard"])
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: this endpoint is not for student dashboard. Please check "Epic" of your ticket before define tag in ednpoint

File: apps/services/handlers.py [hunk 12]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+async def count_student_accommodation(university_id: int, user_request_id: int,  hostel: CountHostelAccommodationIn,
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: 1) count_hostel_accommodation_cost is better
2) delete param user_request_id
3) hostel is the worst name of the variable. Please rename it to data for example

File: apps/services/handlers.py [hunk 13]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+                                      user=Depends(get_current_user)):
+    """
+    **Count Student Accommodation**
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: Rename it

File: apps/services/handlers.py [hunk 14]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+
+    **Path**:
+    - **university_id**: user university id
+    - **user_request_id**: user request id
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: Delete user_request_id

File: apps/services/handlers.py [hunk 15]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+
+    **Input**:
+    - **start_date_accommodation**: start date of user accommodation
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: start date hostel accommodation of student

File: apps/services/handlers.py [hunk 16]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+    - **end_date_accommodation**: end(last) date of user accommodation
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: end date hostel accommodation of student

File: apps/services/handlers.py [hunk 17]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+    - **bed_place_id**: user bed place id
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: delete "user"

File: apps/services/handlers.py [hunk 18]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+
+    **Return**: summa of accommodation, by month price, dates differance and bed place type
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: total cost of hostel accommodation of student, based on ...

File: apps/services/schemas.py [hunk 19]
```diff
@@ -118,3 +118,13 @@ class UserRequestDetailsViewOut(BaseModel):
     date_review: datetime = None
     remark: str = None
     documents: List[Dict[str, str]]
+
+
+class CountHostelAccommodationIn(BaseModel):
+    start_date_accommodation: date = None
+    end_date_accommodation: date = None
+    bed_place_id: int = None
+
```
Comment (anton-serozhechkin) [inline apps/services/schemas.py:141]: add "hostel_id : int"

File: apps/services/handlers.py [hunk 20]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+    """
+    hostel_query = select(hostel_accommodation_view).where(hostel_accommodation_view.c.user_request_id == user_request_id,
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: It's the wrong implementation. Endpoint receives param hostel_id from USER (UI). Review of user request has still not been created. Endpoint will be used while the admin review user request. He can change hostel, bed place, dates of accommodation. That's why you don't need to make this query and execute it to know the month cost of hostel accomodation

File: apps/services/handlers.py [hunk 21]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+                                                           hostel_accommodation_view.c.university_id == university_id,)
+
+    bed_place_query = select(BedPlace).where(BedPlace.bed_place_id == hostel.bed_place_id)
+
+    hostel_info = await database.fetch_one(hostel_query)
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: suffix info is useless

File: apps/services/handlers.py [hunk 22]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+    bed_place_info = await database.fetch_one(bed_place_query)
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: suffix info is useless

File: apps/services/handlers.py [hunk 23]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+
+    return {
+        "data": hostel_info,
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: in "data" key you should return total_hostel_accommodation_cost key with their value

File: apps/services/handlers.py [hunk 24]
```diff
@@ -277,3 +278,38 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+        "message": f"Result summa of student accommodation: "
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: Message shouldn't contain data, only message that "Cost of hostel accommodation of student was counted successfully"

**Review by anton-serozhechkin — CHANGES_REQUESTED** [review]

File: apps/services/services.py [hunk 25]
```diff
@@ -40,3 +40,20 @@ async def create_user_document(**kwargs):
...
+def calculate_dates_different(start_date: date, end_date: date) -> int:
+    if end_date <= start_date:
+        raise ValueError(
+            message="Start date of hostel accommodation can't be more or equal than end date of hostel accommodation"
+        )
+    return -12 * (start_date.year - end_date.year) - start_date.month + end_date.month
+
+
+def get_month_price_by_bed_place(hostel_month_price: float, bed_place_name: str) -> float:
+    return hostel_month_price * float(bed_place_name)
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: convert it to decimal instead of float

File: apps/services/services.py [hunk 26]
```diff
@@ -40,3 +40,20 @@ async def create_user_document(**kwargs):
...
+
+def calculate_dates_different(start_date: date, end_date: date) -> int:
+    if end_date <= start_date:
+        raise ValueError(
+            message="Start date of hostel accommodation can't be more or equal than end date of hostel accommodation"
+        )
+    return -12 * (start_date.year - end_date.year) - start_date.month + end_date.month
+
+
+def get_month_price_by_bed_place(hostel_month_price: float, bed_place_name: str) -> float:
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: decimal instead of float

File: apps/services/services.py [hunk 27]
```diff
@@ -40,3 +40,20 @@ async def create_user_document(**kwargs):
                                         content=content,
                                         user_request_id=kwargs.get("user_request_id"))
     return await database.execute(query)
+
+
+def calculate_dates_different(start_date: date, end_date: date) -> int:
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: calculate_difference_between_dates_in_months

File: apps/services/services.py [hunk 28]
```diff
@@ -40,3 +40,20 @@ async def create_user_document(**kwargs):
...
+            message="Start date of hostel accommodation can't be more or equal than end date of hostel accommodation"
+        )
+    return -12 * (start_date.year - end_date.year) - start_date.month + end_date.month
+
+
+def get_month_price_by_bed_place(hostel_month_price: float, bed_place_name: str) -> float:
+    return hostel_month_price * float(bed_place_name)
+
+
+def calculate_hostel_accommodation_cost(month_price: float, dates_different: int) -> float:
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: 1) rename method to calculate_total_hostel_accommodation_cost
2) months instead of dates_different
3) returned type is decimal, not float
4) please dates_different rename it to month_count

File: apps/services/services.py [hunk 29]
```diff
@@ -40,3 +40,20 @@ async def create_user_document(**kwargs):
...
                                         user_request_id=kwargs.get("user_request_id"))
     return await database.execute(query)
+
+
+def calculate_dates_different(start_date: date, end_date: date) -> int:
+    if end_date <= start_date:
+        raise ValueError(
+            message="Start date of hostel accommodation can't be more or equal than end date of hostel accommodation"
+        )
+    return -12 * (start_date.year - end_date.year) - start_date.month + end_date.month
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: please, separate this mechanism and make it simplier

File: apps/services/schemas.py [hunk 30]
```diff
@@ -118,3 +118,15 @@ class UserRequestDetailsViewOut(BaseModel):
...
+class CountHostelAccommodationCostIn(BaseModel):
+    hostel_id: int
+    start_date_accommodation: date
+    end_date_accommodation: date
+    bed_place_id: int
+
+
+
+class CountHostelAccommodationCostOut(BaseModel):
+    total_hostel_accommodation_cost: float
```
Comment (anton-serozhechkin) [inline apps/services/schemas.py]: decimal instead of float

File: apps/services/handlers.py [hunk 31]
```diff
@@ -277,3 +280,37 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+
+    **Input**:
+    - **hostel_id**: hostel id
+    - **start_date_accommodation**: start date hostel accommodation of student
+    - **end_date_accommodation**: end(last) date hostel accommodation of student
+    - **bed_place_id**: student bed place id
+
+    **Return**: summa of hostel accommodation cost, based on hostel id
+    """
+    query = select(Hostel).where(Hostel.hostel_id == data.hostel_id, Hostel.hostel_id == university_id)
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: why do u use it?
Hostel.hostel_id == university_id
1) it's wrong
2) we don't need to select hostel using university_id

File: apps/services/handlers.py [hunk 32]
```diff
@@ -277,3 +280,37 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
         "data": response,
         "message": "Got request details"
     }
+
+
+@services_router.post("/{university_id}/count-hostel-accommodation-cost/",
+                      name="create_count_hostel_accommodation_cost ",
+                      response_model=JSENDOutSchema[CountHostelAccommodationCostOut],
+                      tags=["Admin dashboard"])
+async def count_hostel_accommodation_cost(university_id: int, data: CountHostelAccommodationCostIn):
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: where is auth for this endpoint?

File: apps/services/handlers.py [hunk 33]
```diff
@@ -277,3 +280,37 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+
+    **Return**: summa of hostel accommodation cost, based on hostel id
+    """
+    query = select(Hostel).where(Hostel.hostel_id == data.hostel_id, Hostel.hostel_id == university_id)
+    hostel = await database.fetch_one(query)
+
+    query = select(BedPlace).where(BedPlace.bed_place_id == data.bed_place_id)
+    bed_place = await database.fetch_one(query)
+
+    dates_different = calculate_dates_different(data.start_date_accommodation, data.end_date_accommodation)
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: rename dates_different to months_difference_between_dates

File: apps/services/handlers.py [hunk 34]
```diff
@@ -277,3 +280,37 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+    **Path**:
+    - **university_id**: user university id
+
+    **Input**:
+    - **hostel_id**: hostel id
+    - **start_date_accommodation**: start date hostel accommodation of student
+    - **end_date_accommodation**: end(last) date hostel accommodation of student
+    - **bed_place_id**: student bed place id
+
+    **Return**: summa of hostel accommodation cost, based on hostel id
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: based not only on hostel_id, pleasee add additional params or delete it

**Review by anton-serozhechkin — CHANGES_REQUESTED** [review]

File: apps/services/services.py [hunk 35]
```diff
@@ -40,3 +42,21 @@ async def create_user_document(**kwargs):
                                         content=content,
                                         user_request_id=kwargs.get("user_request_id"))
     return await database.execute(query)
+
+
+def calculate_difference_between_dates_in_months(end_date: date, start_date: date) -> int:
+    # TODO move it to IN schema, because now it raise, but user see internal server error
```
Comment (anton-serozhechkin) [inline apps/services/services.py]: move logic to in schema

File: apps/services/schemas.py [hunk 36]
```diff
@@ -118,3 +119,14 @@ class UserRequestDetailsViewOut(BaseModel):
...
+
+class CountHostelAccommodationCostIn(BaseModel):
+    hostel_id: int
+    start_date_accommodation: date
+    end_date_accommodation: date
+    bed_place_id: int
+
+
+class CountHostelAccommodationCostOut(BaseModel):
+    total_hostel_accommodation_cost: Decimal
```
Comment (anton-serozhechkin) [inline apps/services/schemas.py:144]: but UI see integer type, please fix it

**Review by anton-serozhechkin — CHANGES_REQUESTED** [review]

File: apps/services/schemas.py (hunk 36 above)
Comment (Kostiantyn-Salnykov) [reply apps/services/schemas.py:144]: @anton-serozhechkin refer to money pattern, monies - only decimal type. ISO-4217 also has details about number of decimal places for every currency.
Design is not a point of truth for everything. 😃

File: apps/services/handlers.py [hunk 37]
```diff
@@ -277,3 +280,41 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+    **Input**:
+    - **hostel_id**: hostel id
+    - **start_date_accommodation**: start date hostel accommodation of student
+    - **end_date_accommodation**: end date hostel accommodation of student
+    - **bed_place_id**: student bed place type
+
+    **Return**: total cost of student hostel accommodation, based on hostel month price, bed place name,
+                start and end dates hostel accommodation
+    """
+    query = select(Hostel).where(Hostel.hostel_id == data.hostel_id)
```
Comment (Kostiantyn-Salnykov) [inline apps/services/handlers.py]: Move this logic to HostelsService (services.py)

File: apps/services/handlers.py [hunk 38]
```diff
@@ -277,3 +280,41 @@ async def read_request_details(university_id: int, user_request_id: int, user=De
...
+    hostel = await database.fetch_one(query)
+
+    query = select(BedPlace).where(BedPlace.bed_place_id == data.bed_place_id)
```
Comment (Kostiantyn-Salnykov) [inline apps/services/handlers.py]: Move this logic to BedPlacesService (services.py)

File: apps/services/schemas.py [hunk 39]
```diff
@@ -122,3 +122,21 @@ class UserRequestDetailsViewOut(BaseModel):
...
     documents: List[Dict[str, str]]
+
+
+class CountHostelAccommodationCostIn(BaseModel):
+    hostel_id: int
+    start_date_accommodation: date
+    end_date_accommodation: date
+    bed_place_id: int
+
+    def __init__(self, **kwargs):
```
Comment (Kostiantyn-Salnykov) [inline apps/services/schemas.py]: This should be implemented via root_validator of Pydantic

File: apps/services/services.py [hunk 40]
```diff
@@ -40,3 +42,16 @@ async def create_user_document(**kwargs):
                                         content=content,
                                         user_request_id=kwargs.get("user_request_id"))
     return await database.execute(query)
+
+
+def calculate_difference_between_dates_in_months(end_date: date, start_date: date) -> int:
```
Comment (Kostiantyn-Salnykov) [inline apps/services/services.py]: This is a business logic -> move to handlers.py

File: apps/services/services.py [hunk 41]
```diff
@@ -40,3 +42,16 @@ async def create_user_document(**kwargs):
...
+    return end_date.month - start_date.month + 12*(end_date.year - start_date.year)
+
+
+def get_month_price_by_bed_place(hostel_month_price: Decimal, bed_place_name: str) -> Decimal:
```
Comment (Kostiantyn-Salnykov) [inline apps/services/services.py]: This is a business logic -> move to handlers.py

File: apps/services/services.py [hunk 42]
```diff
@@ -40,3 +42,16 @@ async def create_user_document(**kwargs):
...
+    return hostel_month_price * Decimal(bed_place_name)
+
+
+def calculate_total_hostel_accommodation_cost(month_price: Decimal, months_count: int) -> Decimal:
```
Comment (Kostiantyn-Salnykov) [inline apps/services/services.py]: This is a business logic -> move to handlers.py

**Review by Kostiantyn-Salnykov — CHANGES_REQUESTED** [review]

**Review by anton-serozhechkin — APPROVED** [review]

**Review by anton-serozhechkin — DISMISSED** [review]

**Review by victor-nuzhniy — CHANGES_REQUESTED** [review]
return await dependant.call(**values)
  File ".\apps\services\routers.py", line 277, in count_hostel_accommodation_cost
    "data": await service_handler.count_hostel_accommodation_cost(
  File ".\apps\services\handlers.py", line 203, in count_hostel_accommodation_cost
    months_count = calculate_difference_between_dates_in_months(data.end_date_accommodation,
NameError: name 'calculate_difference_between_dates_in_months' is not defined

**Review by Kostiantyn-Salnykov — DISMISSED** [review]

**Comment by anton-serozhechkin** [comment]
> ```
> return await dependant.call(**values)
> ```
> 
> File ".\apps\services\routers.py", line 277, in count_hostel_accommodation_cost "data": await service_handler.count_hostel_accommodation_cost( File ".\apps\services\handlers.py", line 203, in count_hostel_accommodation_cost months_count = calculate_difference_between_dates_in_months(data.end_date_accommodation, NameError: name 'calculate_difference_between_dates_in_months' is not defined

**Review by anton-serozhechkin — APPROVED** [review]
